### PR TITLE
Added tinymembench in LKP

### PR DESCRIPTION
**Pull request by SSapkal**
Integrate tinymembench in lkp.

Signed-off-by: Swapnil Sapkal <swapnil.sapkal@amd.com>

### CODE REVIEW

**Comment by SSapkal** [comment]
Results: 

```tinymembench``` Benchmark: 

```
2022-10-04 07:04:44 ./tinymembench
tinymembench v0.4.9 (simple benchmark for memory throughput and latency)

==========================================================================
== Memory bandwidth tests                                               ==
==                                                                      ==
== Note 1: 1MB = 1000000 bytes                                          ==
== Note 2: Results for 'copy' tests show how many bytes can be          ==
==         copied per second (adding together read and writen           ==
==         bytes would have provided twice higher numbers)              ==
== Note 3: 2-pass copy means that we are using a small temporary buffer ==
==         to first fetch data into it, and only then write it to the   ==
==         destination (source -> L1 cache, L1 cache -> destination)    ==
== Note 4: If sample standard deviation exceeds 0.1%, it is shown in    ==
==         brackets                                                     ==
==========================================================================

 C copy backwards                                     :  13452.2 MB/s (1.2%)
 C copy backwards (32 byte blocks)                    :  13441.0 MB/s (1.3%)
 C copy backwards (64 byte blocks)                    :  13386.6 MB/s (0.6%)
 C copy                                               :  13593.6 MB/s (1.5%)
 C copy prefetched (32 bytes step)                    :  13507.9 MB/s (0.9%)
 C copy prefetched (64 bytes step)                    :  13706.0 MB/s (1.2%)
 C 2-pass copy                                        :   9254.1 MB/s
 C 2-pass copy prefetched (32 bytes step)             :   9770.2 MB/s (0.3%)
 C 2-pass copy prefetched (64 bytes step)             :  10103.4 MB/s (0.3%)
 C fill                                               :  24101.6 MB/s (1.9%)
 C fill (shuffle within 16 byte blocks)               :  24085.9 MB/s (1.4%)
 C fill (shuffle within 32 byte blocks)               :  24112.1 MB/s (1.8%)
 C fill (shuffle within 64 byte blocks)               :  22085.0 MB/s
 ---
 standard memcpy                                      :  23862.6 MB/s (0.9%)
 standard memset                                      :  33541.6 MB/s (1.5%)
 ---
 MOVSB copy                                           :  17720.6 MB/s
 MOVSD copy                                           :  17672.8 MB/s (0.5%)
 SSE2 copy                                            :  17668.2 MB/s (0.5%)
 SSE2 nontemporal copy                                :  22965.5 MB/s (2.7%)
 SSE2 copy prefetched (32 bytes step)                 :  17633.3 MB/s (0.4%)
 SSE2 copy prefetched (64 bytes step)                 :  17508.6 MB/s (0.2%)
 SSE2 nontemporal copy prefetched (32 bytes step)     :  19336.2 MB/s (2.1%)
 SSE2 nontemporal copy prefetched (64 bytes step)     :  21257.2 MB/s (2.9%)
 SSE2 2-pass copy                                     :  12626.5 MB/s (0.7%)
 SSE2 2-pass copy prefetched (32 bytes step)          :  12714.8 MB/s (0.9%)
 SSE2 2-pass copy prefetched (64 bytes step)          :  12496.0 MB/s (1.0%)
 SSE2 2-pass nontemporal copy                         :   3745.3 MB/s (0.2%)
 SSE2 fill                                            :  27842.4 MB/s (1.2%)
 SSE2 nontemporal fill                                :  25302.5 MB/s (0.4%)

==========================================================================
== Framebuffer read tests.                                              ==
==                                                                      ==
== Many ARM devices use a part of the system memory as the framebuffer, ==
== typically mapped as uncached but with write-combining enabled.       ==
== Writes to such framebuffers are quite fast, but reads are much       ==
== slower and very sensitive to the alignment and the selection of      ==
== CPU instructions which are used for accessing memory.                ==
==                                                                      ==
== Many x86 systems allocate the framebuffer in the GPU memory,         ==
== accessible for the CPU via a relatively slow PCI-E bus. Moreover,    ==
== PCI-E is asymmetric and handles reads a lot worse than writes.       ==
==                                                                      ==
== If uncached framebuffer reads are reasonably fast (at least 100 MB/s ==
== or preferably >300 MB/s), then using the shadow framebuffer layer    ==
== is not necessary in Xorg DDX drivers, resulting in a nice overall    ==
== performance improvement. For example, the xf86-video-fbturbo DDX     ==
== uses this trick.                                                     ==
==========================================================================

 MOVSD copy (from framebuffer)                        :  27909.6 MB/s
 MOVSD 2-pass copy (from framebuffer)                 :  23270.2 MB/s
 SSE2 copy (from framebuffer)                         :  27924.1 MB/s
 SSE2 2-pass copy (from framebuffer)                  :  22712.1 MB/s

==========================================================================
== Memory latency test                                                  ==
==                                                                      ==
== Average time is measured for random memory accesses in the buffers   ==
== of different sizes. The larger is the buffer, the more significant   ==
== are relative contributions of TLB, L1/L2 cache misses and SDRAM      ==
== accesses. For extremely large buffer sizes we are expecting to see   ==
== page table walk with several requests to SDRAM for almost every      ==
== memory access (though 64MiB is not nearly large enough to experience ==
== this effect to its fullest).                                         ==
==                                                                      ==
== Note 1: All the numbers are representing extra time, which needs to  ==
==         be added to L1 cache latency. The cycle timings for L1 cache ==
==         latency can be usually found in the processor documentation. ==
== Note 2: Dual random read means that we are simultaneously performing ==
==         two independent memory accesses at a time. In the case if    ==
==         the memory subsystem can't handle multiple outstanding       ==
==         requests, dual random read has the same timings as two       ==
==         single reads performed one after another.                    ==
==========================================================================

block size : single random read / dual random read, [MADV_NOHUGEPAGE]
      1024 :    0.0 ns          /     0.0 ns
      2048 :    0.0 ns          /     0.0 ns
      4096 :    0.0 ns          /     0.0 ns
      8192 :    0.0 ns          /     0.0 ns
     16384 :    0.0 ns          /     0.0 ns
     32768 :    0.0 ns          /     0.0 ns
     65536 :    1.3 ns          /     1.9 ns
    131072 :    2.0 ns          /     2.4 ns
    262144 :    2.3 ns          /     2.6 ns
    524288 :    4.1 ns          /     5.0 ns
   1048576 :   10.5 ns          /    13.8 ns
   2097152 :   13.7 ns          /    16.6 ns
   4194304 :   15.2 ns          /    17.5 ns
   8388608 :   16.1 ns          /    18.0 ns
  16777216 :   20.1 ns          /    23.2 ns
  33554432 :   32.3 ns          /    42.7 ns
  67108864 :   75.5 ns          /    97.7 ns

block size : single random read / dual random read, [MADV_HUGEPAGE]
      1024 :    0.0 ns          /     0.0 ns
      2048 :    0.0 ns          /     0.0 ns
      4096 :    0.0 ns          /     0.0 ns
      8192 :    0.0 ns          /     0.0 ns
     16384 :    0.0 ns          /     0.0 ns
     32768 :    0.0 ns          /     0.0 ns
     65536 :    1.3 ns          /     1.9 ns
    131072 :    2.0 ns          /     2.4 ns
    262144 :    2.3 ns          /     2.6 ns
    524288 :    2.8 ns          /     3.2 ns
   1048576 :    8.5 ns          /    11.7 ns
   2097152 :   11.5 ns          /    14.2 ns
   4194304 :   13.1 ns          /    15.0 ns
   8388608 :   13.4 ns          /    15.3 ns
  16777216 :   13.8 ns          /    15.4 ns
  33554432 :   14.5 ns          /    16.5 ns
  67108864 :   61.1 ns          /    85.2 ns
```

```stats.csv``` for Tinymembench:

```
tinymembench.standard_memcpy_MBpS,23862.6
tinymembench.standard_memcpy_MBpS_stddev,0.9
tinymembench.standard_memset_MBpS,33541.6
tinymembench.standard_memset_MBpS_stddev,1.5
tinymembench.MOVSB_copy_MBpS,17720.6
tinymembench.MOVSD_copy_MBpS,17672.8
tinymembench.MOVSD_copy_MBpS_stddev,0.5
tinymembench.SSE2_copy_MBpS,17668.2
tinymembench.SSE2_copy_MBpS_stddev,0.5
tinymembench.SSE2_nontemporal_copy_MBpS,22965.5
tinymembench.SSE2_nontemporal_copy_MBpS_stddev,2.7
tinymembench.SSE2_copy_prefetched_32_bytes_step_MBpS,17633.3
tinymembench.SSE2_copy_prefetched_32_bytes_step_MBpS_stddev,0.4
tinymembench.SSE2_copy_prefetched_64_bytes_step_MBpS,17508.6
tinymembench.SSE2_copy_prefetched_64_bytes_step_MBpS_stddev,0.2
tinymembench.SSE2_nontemporal_copy_prefetched_32_bytes_step_MBpS,19336.2
tinymembench.SSE2_nontemporal_copy_prefetched_32_bytes_step_MBpS_stddev,2.1
tinymembench.SSE2_nontemporal_copy_prefetched_64_bytes_step_MBpS,21257.2
tinymembench.SSE2_nontemporal_copy_prefetched_64_bytes_step_MBpS_stddev,2.9
tinymembench.SSE2_2_pass_copy_MBpS,12626.5
tinymembench.SSE2_2_pass_copy_MBpS_stddev,0.7
tinymembench.SSE2_2_pass_copy_prefetched_32_bytes_step_MBpS,12714.8
tinymembench.SSE2_2_pass_copy_prefetched_32_bytes_step_MBpS_stddev,0.9
tinymembench.SSE2_2_pass_copy_prefetched_64_bytes_step_MBpS,12496.0
tinymembench.SSE2_2_pass_copy_prefetched_64_bytes_step_MBpS_stddev,1.0
tinymembench.SSE2_2_pass_nontemporal_copy_MBpS,3745.3
tinymembench.SSE2_2_pass_nontemporal_copy_MBpS_stddev,0.2
tinymembench.SSE2_fill_MBpS,27842.4
tinymembench.SSE2_fill_MBpS_stddev,1.2
tinymembench.SSE2_nontemporal_fill_MBpS,25302.5
tinymembench.SSE2_nontemporal_fill_MBpS_stddev,0.4
tinymembench.MOVSD_copy_from_framebuffer_MBpS,27909.6
tinymembench.MOVSD_2_pass_copy_from_framebuffer_MBpS,23270.2
tinymembench.SSE2_copy_from_framebuffer_MBpS,27924.1
tinymembench.SSE2_2_pass_copy_from_framebuffer_MBpS,22712.1
tinymembench.single_random_read_65536_block_nohugepage,1.3
tinymembench.dual_random_read_65536_block_nohugepage,1.9
tinymembench.single_random_read_262144_block_nohugepage,2.3
tinymembench.dual_random_read_262144_block_nohugepage,2.6
tinymembench.single_random_read_1048576_block_nohugepage,10.5
tinymembench.dual_random_read_1048576_block_nohugepage,13.8
tinymembench.single_random_read_65536_block_hugepage,1.3
tinymembench.dual_random_read_65536_block_hugepage,1.9
tinymembench.single_random_read_262144_block_hugepage,2.3
tinymembench.dual_random_read_262144_block_hugepage,2.6
tinymembench.single_random_read_1048576_block_hugepage,8.5
tinymembench.dual_random_read_1048576_block_hugepage,11.7
```

```ncompare``` Results:

```
lkp ncompare -s testcase=tinymembench/commit=5.19.12 -s testcase=tinymembench/commit=6.0.0 -a -k
=========================================================================================
compiler/kconfig/rootfs/tbox_group/testcase:
  gcc/defconfig/rhel/souparnika.amd.com/tinymembench

commit:
  5.19.12
  6.0.0

         5.19.12                       6.0.0
---------------- ---------------------------
         %stddev     %change         %stddev
             \          |                \
     20601           -19.2%      16655 ±  4%  tinymembench.MOVSB_copy_MBpS
      0.70           -14.3%       0.60 ± 75%  tinymembench.MOVSB_copy_MBpS_stddev
     25528            -4.7%      24327 ±  3%  tinymembench.MOVSD_2_pass_copy_from_framebuffer_MBpS
      1.60           -95.8%       0.07 ±141%  tinymembench.MOVSD_2_pass_copy_from_framebuffer_MBpS_stddev
     20730           -21.1%      16347 ±  5%  tinymembench.MOVSD_copy_MBpS
      0.60           -33.3%       0.40 ± 35%  tinymembench.MOVSD_copy_MBpS_stddev
     37950           -11.7%      33515 ± 11%  tinymembench.MOVSD_copy_from_framebuffer_MBpS
      0.10            +0.0%       0.10 ± 81%  tinymembench.MOVSD_copy_from_framebuffer_MBpS_stddev
     14078            -7.2%      13063 ±  4%  tinymembench.SSE2_2_pass_copy_MBpS
      0.10          +533.3%       0.63 ± 64%  tinymembench.SSE2_2_pass_copy_MBpS_stddev
     23815            -3.7%      22922        tinymembench.SSE2_2_pass_copy_from_framebuffer_MBpS
      0.20          -100.0%       0.00        tinymembench.SSE2_2_pass_copy_from_framebuffer_MBpS_stddev
     14184            -7.4%      13130 ±  3%  tinymembench.SSE2_2_pass_copy_prefetched_32_bytes_step_MBpS
      0.20          +166.7%       0.53 ± 72%  tinymembench.SSE2_2_pass_copy_prefetched_32_bytes_step_MBpS_stddev
     13713            -6.9%      12761 ±  4%  tinymembench.SSE2_2_pass_copy_prefetched_64_bytes_step_MBpS
      0.50           +80.0%       0.90 ±  9%  tinymembench.SSE2_2_pass_copy_prefetched_64_bytes_step_MBpS_stddev
      3719            +0.5%       3735        tinymembench.SSE2_2_pass_nontemporal_copy_MBpS
      1.00           -36.7%       0.63 ± 96%  tinymembench.SSE2_2_pass_nontemporal_copy_MBpS_stddev
     18903            -3.9%      18163 ±  2%  tinymembench.SSE2_copy_MBpS
      0.00       +3.7e+101%       0.37 ± 34%  tinymembench.SSE2_copy_MBpS_stddev
     38855           -12.5%      33988 ± 12%  tinymembench.SSE2_copy_from_framebuffer_MBpS
      0.10          -100.0%       0.00        tinymembench.SSE2_copy_from_framebuffer_MBpS_stddev
     18798            -3.5%      18134        tinymembench.SSE2_copy_prefetched_32_bytes_step_MBpS
      0.60           -50.0%       0.30 ± 27%  tinymembench.SSE2_copy_prefetched_32_bytes_step_MBpS_stddev
     18625            -3.3%      18016 ±  2%  tinymembench.SSE2_copy_prefetched_64_bytes_step_MBpS
      0.60           -61.1%       0.23 ± 88%  tinymembench.SSE2_copy_prefetched_64_bytes_step_MBpS_stddev
     32772           +13.7%      37251 ± 17%  tinymembench.SSE2_fill_MBpS
      0.00         +6e+101%       0.60 ± 70%  tinymembench.SSE2_fill_MBpS_stddev
     23826            -4.1%      22852 ±  2%  tinymembench.SSE2_nontemporal_copy_MBpS
      0.10         +1600.0%       1.70 ± 59%  tinymembench.SSE2_nontemporal_copy_MBpS_stddev
     24128           -12.5%      21105 ±  9%  tinymembench.SSE2_nontemporal_copy_prefetched_32_bytes_step_MBpS
      0.00       +1.9e+102%       1.90 ± 69%  tinymembench.SSE2_nontemporal_copy_prefetched_32_bytes_step_MBpS_stddev
     24102            -7.0%      22405 ±  4%  tinymembench.SSE2_nontemporal_copy_prefetched_64_bytes_step_MBpS
      0.00       +1.5e+102%       1.53 ± 77%  tinymembench.SSE2_nontemporal_copy_prefetched_64_bytes_step_MBpS_stddev
     25476            -0.5%      25340        tinymembench.SSE2_nontemporal_fill_MBpS
      0.00         +2e+101%       0.20 ± 81%  tinymembench.SSE2_nontemporal_fill_MBpS_stddev
     11.40            +1.5%      11.57        tinymembench.dual_random_read_1048576_block_hugepage
     13.40            +2.0%      13.67        tinymembench.dual_random_read_1048576_block_nohugepage
      2.60            +0.0%       2.60        tinymembench.dual_random_read_262144_block_hugepage
      2.60            +0.0%       2.60        tinymembench.dual_random_read_262144_block_nohugepage
      1.90            -0.0%       1.90        tinymembench.dual_random_read_65536_block_hugepage
      1.90            -0.0%       1.90        tinymembench.dual_random_read_65536_block_nohugepage
      8.30            +1.6%       8.43        tinymembench.single_random_read_1048576_block_hugepage
     10.20            +1.6%      10.37        tinymembench.single_random_read_1048576_block_nohugepage
      2.30            +0.0%       2.30        tinymembench.single_random_read_262144_block_hugepage
      2.30            +0.0%       2.30        tinymembench.single_random_read_262144_block_nohugepage
      1.30            +0.0%       1.30        tinymembench.single_random_read_65536_block_hugepage
      1.30            +0.0%       1.30        tinymembench.single_random_read_65536_block_nohugepage
     24626            -2.6%      23984        tinymembench.standard_memcpy_MBpS
      0.00       +6.3e+101%       0.63 ± 71%  tinymembench.standard_memcpy_MBpS_stddev
     38048           +19.3%      45374 ± 18%  tinymembench.standard_memset_MBpS
      0.80            -8.3%       0.73 ± 74%  tinymembench.standard_memset_MBpS_stddev
```

**Comment by SSapkal** [comment]
Currently after running tinymembench via LKP I am seeing warning messages for empty stats value. This is because tinymembench outputs standard deviation values only if its greater than 0.1%. Opened an issue for the same.

Link: [https://github.com/intel/lkp-tests/issues/176](url)

**Comment by bssrikanth** [comment]
Thanks @SSapkal , If warnings which is pointed in issue #176 are harmless then the changes LGTM. Lets wait for @rli9 comments.

File: stats/tinymembench [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+LKP_SRC = ENV['LKP_SRC'] || File.dirname(File.dirname(File.realpath($PROGRAM_NAME)))
+
+require "#{LKP_SRC}/lib/log"
+
+stats = Hash.new { |h, k| h[k] = [] }
+
+while (line = $stdin.gets)
+  case line
+  when /(standard mem|MOVSB|MOVSD|SSE2)/
```
Comment (rli9) [inline stats/tinymembench:11]: thanks, help comment like # SSE2 copy                                            :  17668.2 MB/s (0.5%)
so easier to know how the line looks like, and what below code tries to parse

File: stats/tinymembench [hunk 2]
```diff
@@ -0,0 +1,43 @@
+#!/usr/bin/env ruby
+
```
Comment (rli9) [inline stats/tinymembench:2]: help add rspec test under spec/stats

File: stats/tinymembench [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    line = line.split(':')
+    key = line[0]
+    key.lstrip!
+    key.rstrip!
+    key = key.gsub(/[)(]/, '')
+    key = key.gsub(/[ -]/, '_')
+    key = key.gsub(/$/, '_MBpS')
+    values = line[1].split(' MB/s')
+    values[0].lstrip!
+    stats[key] = values[0]
```
Comment (rli9) [inline stats/tinymembench:21]: prefer
`stats[key] = values[0].lstrip`

File: stats/tinymembench [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    stats["#{key}_stddev"] = values[1].gsub(/[()% ]/, '') unless values[1].nil?
+  when /(1024|4096|16384|65536|262144|1048576)/
+    line = line.split(':')
+    line[0].lstrip!
+    line[0].rstrip!
+    values = line[1].split('/')
+    if !stats.key?("single_random_read_#{line[0]}_block_nohugepage")
```
Comment (rli9) [inline stats/tinymembench:28]: prefer to define a var like below to be used
`single_key = "single_random_read_#{line[0]}_block_nohugepage"`

File: stats/tinymembench (hunk 1 above)
Comment (SSapkal) [reply stats/tinymembench:11]: Added comment for the same.

File: stats/tinymembench (hunk 2 above)
Comment (SSapkal) [reply stats/tinymembench:2]: Added rspec test under spec/stats for tinymembench.

File: stats/tinymembench [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+while (line = $stdin.gets)
+  case line
+  when /(standard mem|MOVSB|MOVSD|SSE2)/
+    line = line.split(':')
+    key = line[0]
+    key.lstrip!
+    key.rstrip!
+    key = key.gsub(/[)(]/, '')
+    key = key.gsub(/[ -]/, '_')
+    key = key.gsub(/$/, '_MBpS')
```
[the remark this answers is not in the document]
Comment (SSapkal) [reply stats/tinymembench]: Changes done according to the comment.

File: stats/tinymembench (hunk 3 above)
Comment (SSapkal) [reply stats/tinymembench]: Changes done according to the comment.

File: stats/tinymembench (hunk 4 above)
Comment (SSapkal) [reply stats/tinymembench]: Used following vars and appended ```hugepage``` and ```nohugepage``` accordingly
```
single_key = "single_random_read_#{line[0]}_block"
dual_key = "dual_random_read_#{line[0]}_block"
```

**Comment by rli9** [comment]
thanks for the patch

**Pull request merged**